File: packages/sdk/src/core/managers/event-manager.ts
import type { HttpClient } from '../client/http-client';
import type { OfflineQueue } from '../queue/offline-queue';
import type { CreateEventRequest, EventParams } from '../types';
import { logger } from '../utils/logger';
import { EventDeduplicator, RateLimiter } from '../utils/rate-limiter';
import { validateEventName, validateEventParams } from '../utils/validator';

export class EventManager {
  private isOnline = true;
  private readonly httpClient: HttpClient;
  private readonly offlineQueue: OfflineQueue;
  private readonly getSessionId: () => string | null;
  private readonly rateLimiter = new RateLimiter();
  private readonly deduplicator = new EventDeduplicator();

  constructor(
    httpClient: HttpClient,
    offlineQueue: OfflineQueue,
    getSessionId: () => string | null
  ) {
    this.httpClient = httpClient;
    this.offlineQueue = offlineQueue;
    this.getSessionId = getSessionId;
  }

  updateNetworkState(isOnline: boolean): void {
    this.isOnline = isOnline;
  }

  track(name: string, params?: EventParams, isScreen = false): void {
    const nameValidation = validateEventName(name);
    if (!nameValidation.success) {
      logger.error(
        'Invalid event name. Use alphanumeric, _, -, ., /, or space'
      );
      return;
    }

    const paramsValidation = validateEventParams(params);
    if (!paramsValidation.success) {
      logger.error('Invalid event params. Use primitive values only.');
      return;
    }

    if (!this.rateLimiter.canTrack()) {
      return;
    }

    if (this.deduplicator.isDuplicate(name, params)) {
      logger.warn('Duplicate event detected. Ignoring.', { name });
      return;
    }

    const sessionId = this.getSessionId();
    if (!sessionId) {
      logger.error('Session not started. Cannot track event.');
      return;
    }

    const payload: CreateEventRequest = {
      sessionId,
      name,
      params,
      isScreen,
      timestamp: new Date().toISOString(),
    };

    this.sendEvent(payload).catch(() => {
      logger.error('Unhandled error in sendEvent. Event may be lost.');
    });
  }

  private async sendEvent(payload: CreateEventRequest): Promise<void> {
    if (this.isOnline) {
      const result = await this.httpClient.createEvent(payload);
      if (!result.success) {
        logger.error('Event tracking failed. Queuing for retry.', result.error);
        try {
          await this.offlineQueue.enqueue({ type: 'event', payload });
        } catch (error) {
          logger.error('Failed to queue event. Data may be lost.', error);
        }
      }
    } else {
      try {
        await this.offlineQueue.enqueue({ type: 'event', payload });
      } catch (error) {
        logger.error('Failed to queue event. Data may be lost.', error);
      }
    }
  }
}

File: packages/swift/Sources/PhaseAnalytics/Core/Managers/EventManager.swift
import Foundation

internal actor EventManager {
    private var isOnline = true
    private let httpClient: HTTPClient
    private let offlineQueue: OfflineQueue
    private let getSessionID: @Sendable () async -> String?
    private let rateLimiter = RateLimiter()
    private let deduplicator = EventDeduplicator()

    init(
        httpClient: HTTPClient,
        offlineQueue: OfflineQueue,
        getSessionID: @escaping @Sendable () async -> String?
    ) {
        self.httpClient = httpClient
        self.offlineQueue = offlineQueue
        self.getSessionID = getSessionID
    }

    func updateNetworkState(isOnline: Bool) {
        self.isOnline = isOnline
    }

    func track(name: String, params: EventParams?, isScreen: Bool = false) async {
        guard case .success = Validator.validateEventName(name) else {
            logger.error("Invalid event name")
            return
        }

        guard case .success = Validator.validateEventParams(params) else {
            logger.error("Invalid event params")
            return
        }

        guard rateLimiter.canTrack() else {
            return
        }

        guard !deduplicator.isDuplicate(name: name, params: params) else {
            logger.debug("Duplicate event ignored: \(name)")
            return
        }

        guard let sessionID = await getSessionID() else {
            logger.error("Session not started, cannot track event")
            return
        }

        let convertedParams: [String: AnyCodable]?
        if let params = params {
            convertedParams = params.dictionary.mapValues { AnyCodable($0) }
        } else {
            convertedParams = nil
        }

        let payload = CreateEventRequest(
            sessionId: sessionID,
            name: name,
            params: convertedParams,
            isScreen: isScreen,
            timestamp: currentISO8601Timestamp()
        )

        await sendEvent(payload)
    }

    private func sendEvent(_ payload: CreateEventRequest) async {
        if isOnline {
            let result = await httpClient.createEvent(payload)
            if case .failure(let error) = result {
                logger.error("Failed to track event, queueing", error)
                await offlineQueue.enqueue(.event(payload: payload, clientOrder: 0, retryCount: nil))
            }
        } else {
            await offlineQueue.enqueue(.event(payload: payload, clientOrder: 0, retryCount: nil))
        }
    }
}
